client/src/components/CustomFields.js

```diff
@@ -633,51 +633,75 @@ const FIELD_COMPONENTS = {
+  // loop through fields of the config to check for visibility of a field
+  Object.entries(fieldsConfig).forEach(([key, fieldConfig]) => {
+    const visibleWhenPath = fieldConfig.visibleWhen
     const isVisible =
-      !fieldConfig.visibleWhen ||
-      (fieldConfig.visibleWhen &&
-        !_isEmpty(JSONPath(fieldConfig.visibleWhen, formikValues)))
-    if (!isVisible && !prevInvisibleFields.includes(fieldName)) {
-      turnedInvisible.push(fieldName)
-    } else if (isVisible && prevInvisibleFields.includes(fieldName)) {
-      turnedVisible.push(fieldName)
+      !visibleWhenPath || !_isEmpty(JSONPath(visibleWhenPath, formikValues))
+
+    const fieldName = `${parentFieldName}.${key}`
+
+    // recursively append invisible fields in case of array of objects
+    // we can have customFields.array_of_objects.objectFields.array_of_objects.objectFields...
+    if (fieldConfig.type === CUSTOM_FIELD_TYPE.ARRAY_OF_OBJECTS) {
+      curInvisibleFields.push(
+        ...getInvisibleFields(
+          fieldConfig.objectFields,
+          fieldName,
+          formikValues,
+          true
+        )
+      )
+    }
+    if (!isVisible) {
+      if (isArrayOfObjects) {
+        // insert index to fieldName
+        // so that we can acces array items with _set, _get methods
+        _get(formikValues, parentFieldName).forEach((obj, index) => {
+          curInvisibleFields.push(`${parentFieldName}.${index}.${key}`)
+        })
+      } else {
+        curInvisibleFields.push(fieldName)
+      }
+    }
+  })
+  return curInvisibleFields
+}
+
+function resetInvalidInvisibleFields(invisibleFields, setFieldValue, errors) {
+  invisibleFields?.forEach(fieldName => {
+    if (_get(errors, fieldName)) {
+      setFieldValue(fieldName, null)
     }
   })
-  if (turnedVisible.length || turnedInvisible.length) {
-    curInvisibleFields = prevInvisibleFields.filter(
-      x => !turnedVisible.includes(x)
-    )
-    turnedInvisible.forEach(x => curInvisibleFields.push(x))
-    return curInvisibleFields
-  }
-  return invisibleFields
 }
 
 export const CustomFieldsContainer = props => {
-  const { parentFieldName, formikProps } = props
-  const [invisibleFields, setInvisibleFields] = useState([])
-  const prevInvisibleFields = useRef(invisibleFields)
-  const { setFieldValue } = formikProps
+  const {
+    parentFieldName,
+    formikProps: { values, errors, setFieldValue },
+    fieldsConfig
+  } = props
+
+  const invisibleFields = useMemo(
+    () => getInvisibleFields(fieldsConfig, parentFieldName, values),
+    [fieldsConfig, parentFieldName, values]
+  )
+
   const invisibleFieldsFieldName = `${parentFieldName}.${INVISIBLE_CUSTOM_FIELDS_FIELD}`
   useEffect(() => {
-    if (!_isEqual(invisibleFields, prevInvisibleFields.current)) {
-      prevInvisibleFields.current = invisibleFields
+    if (!_isEqual(_get(values, invisibleFieldsFieldName), invisibleFields)) {
+      resetInvalidInvisibleFields(invisibleFields, setFieldValue, errors)
```

Comment:
This is actually new functionality, I think it didn't use to reset the value when invalid.
Is this on purpose?

Reply:
Yes, it is a new functionality, i somehow thought that we should reset if the value is invalid when invisible. Maybe this way is better or worse. @VassilIordanov do you have preference?

Reply:
If the user filled something in and somehow clicked on something by mistake making the field invisible when making it visible again the data he/she filled in is lost. And it might be a lot of data, or the user might not even realize that the data of one of the fields was removed.
I would think that it's better if the user sees the validation error and gets the change to fix it.